lib/improver/spotdata/extrema.py

```diff
@@ -0,0 +1,249 @@
+
+        """
+        cube.coord('time').points = cube.coord('time').points.astype(np.int64)
+
+        # Adjust times on cube to be local to each site.
+        local_tz_cube = make_local_time_cube(cube)
+
+        # Starts at start_hour on first available day, runs until start_hour on
+        # final_date.
+        start_time, end_time = get_valid_dates(local_tz_cube.coord('time'),
+                                               self.start_hour)
+        num_periods = int(
+            np.ceil((end_time - start_time).total_seconds()/3600/self.period))
+        starts = [start_time + datetime.timedelta(hours=i*self.period)
+                  for i in range(num_periods)]
+        ends = [time + datetime.timedelta(hours=self.period)
+                for time in starts]
+
+        # Extract extrema values over desired time periods, producing a cube
+        # for each period.
+        period_cubes = CubeList()
+        for period_start, period_end in zip(starts, ends):
+            extrema_constraint = datetime_constraint(period_start, period_end)
+            with iris.FUTURE.context(cell_datetime_objects=True):
+                cube_over_period = local_tz_cube.extract(extrema_constraint)
+            if cube_over_period is not None:
+                # Ensure time dimension of resulting cube reflects period.
+                mid_time = dt_to_utc_hours(period_start +
+                                           (period_end - period_start)/2)
+                bounds = [dt_to_utc_hours(period_start),
+                          dt_to_utc_hours(period_end)]
+
+                extremas = [['max', IMAX], ['min', IMIN]]
+                for name, method in extremas:
+                    cube_out = cube_over_period.collapsed('time', method)
+                    cube_out.long_name = cube_out.name() + '_' + name
+                    cube_out.standard_name = None
+                    cube_out.coord('time').convert_units(
+                        'hours since 1970-01-01 00:00:00')
+                    cube_out.coord('time').points = mid_time
+                    cube_out.coord('time').bounds = bounds
+                    period_cubes.append(cube_out)
+
+        return period_cubes
+
+
+def make_local_time_cube(cube):
+    """
+    Construct a cube in which data is arranged along a dimension coordinate of
+    local time. This allows for the calculation of maxima/minima values over
+    given ranges of local time (e.g. 09Z - 21Z maxima).
+
+    e.g.
+
+    UTC Coord  :  12   13   14   15   16
+    Data: Site 1  300  302  296  294  290 (UTC offset = -2)
+          Site 2  280  282  283  280  279 (UTC offset = +2)
+
+    Data redistributed according to UTC offset to sit on local time.
+
+    Local times:  10   11   12   13   14   15   16   17   18
+    Data: Site 1  300  302  296  294  290  -    -    -    -
+          Site 2  -    -    -    -    280  282  283  280  279
+
+    There will be MDI data in locations which are not forecast at the given
```

Comment:
Is it correct to say `MDI data`? As far as I can tell, invalid values (e.g. NaN, Inf, None) will be masked. It might be ok to just expand this point e.g. `There will be data with a missing data indicator e.g. None in locations...`.